applicationset/generators/generator_spec_processor_test.go

```diff
@@ -0,0 +1,300 @@
+				{"path": "app1", "path.basename": "app1", "path.basenameNormalized": "app1"},
+				{"path": "app2", "path.basename": "app2", "path.basenameNormalized": "app2"},
+				{"path": "app_3", "path.basename": "app_3", "path.basenameNormalized": "app-3"},
+			},
+			expectedError: nil,
+		},
+		{
+			name:        "It filters application according to the paths",
+			directories: []argoprojiov1alpha1.GitDirectoryGeneratorItem{{Path: "p1/*"}, {Path: "p1/*/*"}},
+			repoApps: []string{
+				"app1",
+				"p1/app2",
+				"p1/p2/app3",
+				"p1/p2/p3/app4",
+			},
+			repoError: nil,
+			expected: []map[string]string{
+				{"path": "p1/app2", "path.basename": "app2", "path[0]": "p1", "path.basenameNormalized": "app2"},
+				{"path": "p1/p2/app3", "path.basename": "app3", "path[0]": "p1", "path[1]": "p2", "path.basenameNormalized": "app3"},
+			},
+			expectedError: nil,
+		},
+		{
+			name:        "It filters application according to the paths with Exclude",
+			directories: []argoprojiov1alpha1.GitDirectoryGeneratorItem{{Path: "p1/*", Exclude: true}, {Path: "*"}, {Path: "*/*"}},
+			repoApps: []string{
+				"app1",
+				"app2",
+				"p1/app2",
+				"p1/app3",
+				"p2/app3",
+			},
+			repoError: nil,
+			expected: []map[string]string{
+				{"path": "app1", "path.basename": "app1", "path.basenameNormalized": "app1"},
+				{"path": "app2", "path.basename": "app2", "path.basenameNormalized": "app2"},
+				{"path": "p2/app3", "path.basename": "app3", "path[0]": "p2", "path.basenameNormalized": "app3"},
+			},
+			expectedError: nil,
+		},
+		{
+			name:        "Expecting same exclude behavior with different order",
+			directories: []argoprojiov1alpha1.GitDirectoryGeneratorItem{{Path: "*"}, {Path: "*/*"}, {Path: "p1/*", Exclude: true}},
+			repoApps: []string{
+				"app1",
+				"app2",
+				"p1/app2",
+				"p1/app3",
+				"p2/app3",
+			},
+			repoError: nil,
+			expected: []map[string]string{
+				{"path": "app1", "path.basename": "app1", "path.basenameNormalized": "app1"},
+				{"path": "app2", "path.basename": "app2", "path.basenameNormalized": "app2"},
+				{"path": "p2/app3", "path.basename": "app3", "path[0]": "p2", "path.basenameNormalized": "app3"},
+			},
+			expectedError: nil,
+		},
+		{
+			name:          "handles empty response from repo server",
+			directories:   []argoprojiov1alpha1.GitDirectoryGeneratorItem{{Path: "*"}},
+			repoApps:      []string{},
+			repoError:     nil,
+			expected:      []map[string]string{},
+			expectedError: nil,
+		},
+		{
+			name:          "handles error from repo server",
+			directories:   []argoprojiov1alpha1.GitDirectoryGeneratorItem{{Path: "*"}},
+			repoApps:      []string{},
+			repoError:     fmt.Errorf("error"),
+			expected:      []map[string]string{},
+			expectedError: fmt.Errorf("error"),
+		},
+	}
+	argoCDServiceMock := argoCDServiceMock{mock: &mock.Mock{}}
+
+	argoCDServiceMock.mock.On("GetDirectories", mock.Anything, mock.Anything, mock.Anything).Return(cases[0].repoApps, cases[0].repoError)
```

Comment:
You define several `cases` but only appear to use the first one.

Reply:
Ahh yeah, this was more for testing the `GetRelevantGenerators` function (not mine). I'm fairly new to Go (I'm more on the ArgoCD Yaml side :), so I was using some tests that were used for the Git Generators (where they were specifically testing the Git Generator).

My test cases were below in the `TestGetRelevantGenerators` function:

<img width="911" alt="image" src="https://user-images.githubusercontent.com/29113963/164109474-673c8a96-7665-4fda-8ee5-7899178dda5d.png">

Anyways, I cleaned up the above function you mentioned. Thanks for pointing it out!